docs: add mm bot documentation (#144)
yetanotherco/yet-another-bridge: mm-bot/docs/README.md

# MM Bot
MM Bot is a process designed to supply liquidity to YAB Escrow orders.

## Logical View
### Functional Requirements
- The bot must be able to read an order from the Escrow contract.
- The bot must be able to perform a transfer in Ethereum to the recipient address through the Payment
Registry contract.

> - The bot must be able to perform a repay in Ethereum to recover the funds in the L2 through the

Reviewer: esto es sobre pasar los fondos que recibió en L2 devuelta a L1 ?
no me queda claro a que te referis con "perform a repay"

Payment Registry contract.
- The bot must be able to store the orders in a database and update their status.

> - In case of an error, the bot must be able to store the error and retry the order.

Reviewer: I would add the functional requirement of the bot being able to continue its process where it left off if it has been disconnected for some time

### Simplified Class Diagram
The following diagram shows the mm classes and how they interact with each other.

> ![mm_diagram_class.svg](images%2Fmm_diagram_class.svg)

Reviewer: I this diagram needs a bit of resizing before it can be totally read comfortably

### Full Class Diagram
The following diagram is a detailed version of the previous diagram, 
showing the attributes and methods of each class.

> ![mm_diagram_class_full.svg](images%2Fmm_diagram_class_full.svg)

Reviewer: this diagram can't be appreciated in the context where it will be read by external users (which is directly from github)

## Process View
### Non-Functional Requirements
- The bot must be able to handle multiple orders simultaneously.

> - The bot must be able to retrieve the status of the orders in case of interruption and complete it.

Reviewer: ```suggestion
- The bot must be able to retrieve the status of the orders in case of interruption and complete them.
```

- The bot must be highly available.

> - The bot must index the orders that belong to accepted blocks to ensure that orders are not lost.

Reviewer: ```suggestion
- The bot must index the orders that belong to accepted blocks to ensure the block being read is final and will not change
```

Author: The bot can not revert transactions if the block is reverted

- The bot must be able to retry failed orders.

> - The bot must be able to perform adequate logs for the orders tracking.

Reviewer: ```suggestion
- The bot must generate adequate logs for order tracking.
```

### Architecture
The bot architecture is as follows:

> ![architecture.png](images%2Farchitecture.png)

Reviewer: I think the watches inside the main processes would go better in the top right corner and if possible with some short text saying how frequent ill the process run

The bot is composed of the following components:

> - **MM Bot**: The main process of the bot. It has the following subcomponents:

Reviewer: ```suggestion
- **Main Process**: The main process of the bot. It has the following subcomponents:
```
I wouldnt call the bots main process "MM bot". Also this MM bot is not present on the diagram, Main Process is

>   - `Main Order Indexer`: The `Main Order Indexer` is responsible for indexing the orders from 
>     the pending blocks.
>   - `Order Processor`: The `Order Processor` is responsible for processing the orders.
>   - `Failed Orders Processor`: The `Failed Orders Processor` is responsible for retrying the failed orders.
>   It runs every 5 minutes.
>   - `Accepted Blocks Processor`: The `Accepted Blocks Processor` is responsible for indexing the orders
>     that belong to accepted blocks. It runs every 5 minutes.

Reviewer: ```suggestion
  - `Main Order Indexer`: The `Main Order Indexer` is responsible for indexing the orders from 
    the pending blocks.
  - `Order Processor`: Responsible for processing the orders.
  - `Failed Orders Processor`: Responsible for retrying the failed orders.
  It runs every 5 minutes.
  - `Accepted Blocks Processor`: Responsible for indexing the orders that belong to accepted blocks. It runs every 5 minutes.
```

- **Database**: The database is used to store the following data:

>   - Orders.
>   - Errors.
>   - Block numbers.

Reviewer: ```suggestion
  - Orders
  - Errors
  - Block numbers
```

> An important aspect is that the bot must be able to handle multiple orders simultaneously.
> For that reason, the bot uses asyncio to handle the orders concurrently. This way is preferred
> over using threads, due to the potentially high number of orders that the bot must handle and 
> the fact that the bot is I/O bound.

Reviewer: ```suggestion
An important aspect is that the bot must be able to handle multiple orders simultaneously.
For that reason, the bot uses the library 'asyncio' to handle orders concurrently. This way is preferred
over using threads, due to the potentially high number of orders that the bot must handle and 
the fact that the bot is I/O bound.
```

Reviewer: also, this requirement should be in the "non functional requirements" list

Author: Handle orders concurrently is already in non func req. Using asyncio is an implementatin detail

> Another important aspect is that the bot must have a reliable network connection to communicate
> with Ethereum and L2 networks RPCs.

Reviewer: ```suggestion
Another important requirement is that the bot must have a reliable network connection to communicate
with Ethereum's and L2 networks' RPCs.
```

## Physical View

> The system is deployed in an EC2 virtual machine in AWS.
> The EC2 runs the bot process and the database.

Reviewer: ```suggestion
The system is currently deployed in an EC2 virtual machine in AWS.
This EC2 runs the bot Main Process as well as its database.
```

> ![physical_view.png](images/physical_view.png)

Reviewer: I would add "internet" or a similar label to the vertical dotted line, as well as some label to one or both blockchains.
Otherwise, this diagram does not encompass the whole physical view of the system.

## Scenarios
### 1. Order Processing Flow
The following diagram shows the complete process of an order.

> ![order_processing.svg](images%2Forder_processing.svg)

Reviewer: "Processing order" should be "process order" to match the conjugation of the rest of the verbs in the diagram

Reviewer: Also, "repay funds" is not accurate. That process is as follows:
MM asks PaymentRegistry to withdraw, then if appropiate, PaymentRegistry makes a call to Escrow, and if appropiate, Escrow sends the funds to MM in Starknet

> The order has the following states:

Reviewer: ```suggestion
And each order has the following states:
```

> ![state_diagram.svg](images%2Fstate_diagram.svg)

Reviewer: - starting point should have some sort of label. for example "new order created in escrow", or "new order detected".
- the state labels have an extra row which adds no value, a simple label is enough.
- change "MM starts procesing order" to "Main Process begins"
- Bifurcating paths are correctly noted (with the rhombus), but according to BPMN notation, the question must go inside the rhombus and each arrow coming from it must contain a small text noting the answer to the question. This makes the flow easier to follow.

Author: | change "MM starts procesing order" to "Main Process begins" 
Main process has already started

### 2. Failed Orders Reprocessing

> When an order fails, the bot stores the error and marks the order as failed. So, the `Failed
> Orders Processor` will retry the failed orders. The following diagram shows the flow of a 
> failed order through the bot.

Reviewer: ```suggestion
When an order fails, the bot stores the error, and marks the order as failed. This way, the `Failed
Orders Processor` is able to retry the failed orders. The following diagram shows the flow of a 
failed order through the bot.
```

> ![failed_orders.svg](images%2Ffailed_orders.svg)

Reviewer: - change "store order as failed do to AN error"
- same suggestions as in the previous diagram

### 3. Shutdown Recovery
When the bot starts, it retrieves incomplete orders from the database and continues their processing.

### Accepted Blocks Indexation

> The Main Order Indexer process orders from pending blocks. The `Orders from
> Accepted Blocks Processor` will index the orders that belong to accepted blocks. This way, if the `Main Order
> Indexer` loses an order, it will be taken from the `Orders from Accepted Blocks Processor`.

Reviewer: ```suggestion
The Main Order Indexer processes orders from pending blocks. The `Orders from
Accepted Blocks Processor` will index the orders that belong to accepted blocks. This way, if the `Main Order
Indexer` loses an order, it will be captured and processed by the `Orders from Accepted Blocks Processor`.
```

> ![accepted_blocks.svg](images%2Faccepted_blocks.svg)

Reviewer: Same suggestion as the previous 2 diagrams
